src/aap_eda/wsapi/consumers.py

```diff
@@ -531,39 +537,40 @@ def get_file_contents_from_credentials(
     def get_env_vars_from_credentials(
         self, activation: models.Activation
     ) -> tp.Optional[str]:
-        vault_password, vault_id = self.get_vault_password_and_id(activation)
-        env_vars = {}
-
-        for eda_credential in activation.eda_credentials.all():
-            schema_inputs = eda_credential.credential_type.inputs
-            secret_fields = get_secret_fields(schema_inputs)
-            injectors = eda_credential.credential_type.injectors
-            user_inputs = yaml.safe_load(
-                eda_credential.inputs.get_secret_value()
+        try:
+            vault_password, vault_id = self.get_vault_password_and_id(
+                activation
             )
-            if "env" not in injectors:
-                continue
-
-            if secret_fields:
-                self.encrypt_user_inputs(
-                    secret_fields=secret_fields,
-                    user_inputs=user_inputs,
-                    password=vault_password,
-                    vault_id=vault_id,
-                )
+            env_vars = {}
 
-            for key, value in injectors["env"].items():
-                if key in env_vars:
-                    raise DuplicateEnvKeyError(f"env {key} already exists")
-                env_vars[key] = (
-                    value
-                    if not isinstance(value, str) or "eda.filename" in value
-                    else substitute_variables(value, user_inputs)
+            for eda_credential in activation.eda_credentials.all():
+                schema_inputs = eda_credential.credential_type.inputs
+                secret_fields = get_secret_fields(schema_inputs)
+                injectors = eda_credential.credential_type.injectors
+                user_inputs = yaml.safe_load(
+                    eda_credential.inputs.get_secret_value()
                 )
+                if "env" not in injectors:
+                    continue
 
-        if not env_vars:
-            return None
-        return yaml.dump(env_vars)
+                add_default_values_to_user_inputs(schema_inputs, user_inputs)
```

Comment:
Adding a default value unconditionally would not break any logic that depends on the existence of the variable? For example, it is not the same if you inject a variable for a token with an empty value rather than not injecting it at all. Since the input has not a "default" value in the schema, I think it is assumed that if is not defined, it should not exist, unless it is defined as a required field, in which case it should have a value. 

I feel for the builtin credential types we should handle the absence of not required fields and not pass any undefined field that is not required for builtin or custom types.

Reply:
@Alex-Izquierdo The problem is not that the problem is in the injectors be it file or env the user can access one of these keys for which the user has not provided a value. There is a single credential type called RH AAP which has userid/password, userid/token just token or nothing just the host URL only the host field is required. The other fields might be there or missing but the injectors are built to access all these variables. We can't change the injectors so we have to provide defaults.

Reply:
@mkanoor I added a check on the result of the `substitute_variables` function and removed undefined fields if it returns. The absence of not-required fields now is skipped.

Reply:
@hsong-rh @Alex-Izquierdo I tried this by checking on AAP and we have to send all the env vars that are listed in the injectors down. Its possible for customers when they write their own credential types to provide defaults on a field by field basis which we should honor.